lightning/src/ln/channelmanager.rs

```diff
@@ -691,6 +684,8 @@ pub type SimpleRefChannelManager<'a, 'b, 'c, 'd, 'e, M, T, F, L> = ChannelManage
 //  |       |
 //  |       |__`pending_inbound_payments`
 //  |           |
+//  |           |__`claimable_htlcs`
```

Comment:
From my understanding this lock order tree is currently prescriptive (where we would like to go) but not descriptive (where we're today), in the sense `claimable_htlcs` lock is currently taken in `process_pending_htlc_forwards()` without taking `per_peer_state`, or what do you aim to convey?

Reply:
So the tree is "descriptive" today in the sense that none of the illustration is violated in any of our code (or well after this PR we will violate the branch illustration, but just in the `ChannelManager::write` function).
But you are correct that the `claimable_htlcs` lock for example is never acquired at the same time as the `per_peer_state` lock in our codebase.

However the tree illustration is also ment to give you guidance on which order you are supposed to take the locks in, in case you implement new code in the future that needs to acquire both locks at the same time.
I deliberately choose to illustrate the `per_peer_state` before `pending_inbound_payments` (and therefore also `claimable_htlcs`) even though that's not actually enforced currently by the "actual" lock order, as `pending_inbound_payments` has a "real" lock order currently after `channel_state` currently. And I wanted to avoid that new PRs before #1507 adds code that creates "real" lock orders for `pending_inbound_payments` and `pending_inbound_payments` before `per_peer_state` but after `channel_state` as it would need to be changed once channels are moved to `per_peer_state` with #1507.

Though I've realized now that most of the places in the code that requires that `pending_inbound_payments` is locked before `channel_state` is actually removed with this PR :). I actually created a draft PR #1773 that moves the  `pending_inbound_payments` and `pending_outbound_payments` lock order to before `channel_state`.

Now that I think about it again, we should actually include that lock order change with this PR, as doing so will solve most of the issues I've faced in this PR with the `ChannelManager::write` function, and will allow us to take the `pending_inbound_payments`, `pending_outbound_payments` and `claimable_htlcs` locks and hold them throughout the function 🎉